### ISSUE TITLE
Undefined array key after upgrade to update version 53

### ISSUE THREAD

**Issue by TineUser** [dokuwiki/dokuwiki#3796]
Hi,

I just upgraded our system from PHP7.4.21 to PHP 8.0.19 and DokuWiki to update version 53. I hope this is the correct version information of DokuWiki. I don't know the exact build name. I found this on the administration start page. After upgrading I am getting the following warnings:
```
Warning: Undefined array key "REMOTE_USER" in C:\xampp-dev\htdocs\dokuwiki\lib\plugins\authad\adLDAP\adLDAP.php on line 654

Warning: Undefined array key "mail" in C:\xampp-dev\htdocs\dokuwiki\lib\plugins\authad\auth.php on line 224

Warning: Trying to access array offset on value of type null in C:\xampp-dev\htdocs\dokuwiki\lib\plugins\authad\auth.php on line 224
```
We are only using the AD authentication with the authad plugin. The other authentication plugins are deactivated:

![image](https://user-images.githubusercontent.com/77736099/195016456-4ec1b481-aa00-4e9b-8650-6a16bd292615.png)

The login is working although there warning messages.

Kind Regards

**Comment by mprins** [comment]
It's unclear what version you have, please look at the admin page.
Some of these warnings have been fixed in the master branch, see https://github.com/splitbrain/dokuwiki/issues?q=is%3Aissue+authad+is%3Aclosed

**Comment by TineUser** [comment]
As written: update version 53. I hope this is the correct version information of DokuWiki. I don't know the exact build name. **I found this on the administration start page.**

I downloaded yesterday from the master branch and have still these warnings.

**Comment by Klap-in** [comment]
[NB: @fiwswe please do not recommend enabling phpok. phpok should not be used on normal circumstances, I think it is not worth to enable it for such a test, and if you still recommend it, please explain that it should be _directly_ disabled after the test. If phpok is enabled on a public wiki, or with loggedin users which are not trusted users (=! not admin), the users can retrieve **all** data from that webserver as if they had commandline access ]

By purpose, `$_SERVER['REMOTE_USER']` is not set/null if an user is not logged in. So the global class $INPUT, with `$INPUT->server->str('REMOTE_USER')` function should be used here, which does all the checking (and cleaning) for us, if we request the variable. I think it is worth to use the global $INPUT, even while this library is an external library. 

Formally, it should be fixed in the upstream of this library, it might be worth checking if there are still updates for this (older) library. But there are long term ideas (if I say it right) to replace authad by https://www.dokuwiki.org/plugin:pureldap.

**Comment by TineUser** [comment]
As already written I got the version on the administration start page. Here is the screenshot and you see the version I wrote already:

![image](https://user-images.githubusercontent.com/77736099/195196249-c79d25fa-f5f9-464f-a99d-2a1b0353b820.png)

**Comment by TineUser** [comment]
Instead of using authAD I switched now to pureLDAP. Now I don't have the warning messages any more but when logging in with an administrative account I don't have the settings button to access the administration pages.

Where do I have to configure the AD user group to set the administrator accounts when using pureLDAP for authentication?

**Comment by Klap-in** [comment]
I think you search the superuser config setting https://www.dokuwiki.org/config

To see the groups of your current logged in user, you could add ?do=check (or &do=check) to the url.

**Comment by TineUser** [comment]
I have already set the configuration of the superuser with the AD group name and a leading "@". With authAD this working fine, with pureLDAP not.

I didn't know the URL parameter `do=check`. This helped me to see that pureLDAPs listing of the groups for the user changed in comparison to the authADs listing and the administrative group isn't listed. I increased the parameter plugin»pureldap»page_size to 500 (default is 150) but this didn't help.

**Comment by fiwswe** [comment]
> [NB: @fiwswe please do not recommend enabling phpok. phpok should not be used on normal circumstances, I think it is not worth to enable it for such a test, and if you still recommend it, please explain that it should be directly disabled after the test. If phpok is enabled on a public wiki, or with loggedin users which are not trusted users (=! not admin), the users can retrieve all data from that webserver as if they had commandline access ]

Well, I’m sorry for mentioning a potentially dangerous DW feature in a context meant for developers and admins. As a DW admin, which someone able to change this setting must be, you have lots of potential to shoot yourself in the foot. And the setting is even marked as impacting security on the config page. So I think we can expect the readers to think before making any changes.

Also I think it was clear from the context that this was meant for a test, i.e. to figure out the contents of `$_SERVER` on a particular installation, not as permanent setting.

I do use this on my own wikis (and I am very happy this feature exists) but they are not public and only knowledgeable and trusted users are allowed. I don’t think it was mentioned whether the wiki in question was public or not.

And I agree that phpok should not generally be used on wikis that have untrusted or clueless users with editing permissions.

But all of this is OT for #3796 anyway so I’ll shut up now.

**Comment by TineUser** [comment]
Yes, there's no VERSION file present. The ZIP file I downloaded from the master branch on 10.10.2022 doesn't contain such a file.

**Comment by splitbrain** [comment]
@TineUser since you're running master, please update to the most recent master. Then report back if (and which) errors remain.

**Comment by TineUser** [comment]
I updatet now with the follwing master: 13.10.2022, 08:50 (file date).

I get the following warnings:
```
Warning: Undefined array key "REMOTE_USER" in C:\xampp-dev\htdocs\dokuwiki\lib\plugins\authad\adLDAP\adLDAP.php on line 654
Warning: Undefined array key "mail" in C:\xampp-dev\htdocs\dokuwiki\lib\plugins\authad\auth.php on line 224
Warning: Trying to access array offset on value of type null in C:\xampp-dev\htdocs\dokuwiki\lib\plugins\authad\auth.php on line 224
```
I can log in but I always get these messages.

**Comment by Klap-in** [comment]
master is the development version, which is Igor plus all the additions/changes since August.
The master has no VERSION file.

First warning can be solved by using global $INPUT class, second when should be checked if set. 
First guess is `?? ''` could be enough for the 2nd and 3rd warning together. 

However, it means that your setup does not return a 'mail' field? I do not know if that is common. Could it mean that the other columns can be undefined sometimes as well?

**Comment by TineUser** [comment]
OK, thank you. Yes, not all users have a mail address. When logging in with a normal user that has a mail address the warnings are this ones:
```
Warning: Undefined array key "mail" in C:\xampp-dev\htdocs\dokuwiki\lib\plugins\authad\auth.php on line 224
Warning: Trying to access array offset on value of type null in C:\xampp-dev\htdocs\dokuwiki\lib\plugins\authad\auth.php on line 224
```
I don't know what you mean by 

> First warning can be solved by using global $INPUT class, second when should be checked if set.
First guess is ?? '' could be enough for the 2nd and 3rd warning together.

What can I do?

**Comment by TineUser** [comment]
Is there something I can do for helping to solve the issue?

**Comment by Klap-in** [comment]
If you know php you could open a pull request with the fixes, otherwise others can do that.

**Comment by TineUser** [comment]
No, I'm not a developer. So I'm not able to do this.

**Comment by dipohl** [comment]
I get the same warning message running LDAP Auth Plugin under PHP 8.0
when user data is fetched where no mail address is existent.

```
Warning: Undefined array key "mail" in
/var/www/html/dokuwiki/lib/plugins/authldap/auth.php on line 240

Warning: Trying to access array offset on value of type null in
/var/www/html/dokuwiki/lib/plugins/authldap/auth.php on line 240

```
Release 2022-07-31a "Igor"

LDAP Auth Plugin Version  2015-07-13

**Comment by onkobu** [comment]
Still there with

```
Release 2025-05-14a "Librarian"
PHP 8.2.28
Gentoo Linux
```

A simple appendix ?? null in line 224 of lib/plugins/authldap/auth.php does the trick. (Not all users have an e-mail).
